### ISSUE TITLE
No longer possible to have OpenVPN listen on same port for TCP6/TCP4

### ISSUE THREAD

**Issue by sjorge** [opnsense/core#2422]
I have 4 OpenVPN server instances listening on the same interface on the same port but with different protocols TCP6, TCP4, UDP4, and UDP6. This is working fine. But I am no longer able to update the servers because on save it now complains that the local port is already in use.

The error message is:
```
The following input errors were detected:
    The specified 'Local port' is in use. Please select another value
```

This used to work, but I have not update my config in a while so I do not know when it broke.

![screenshot](https://i.imgur.com/AQMbGvW.jpg)

As you can see this is working fine, I can just no longer edit any of them.

```
sjorge@hydrogen:~ % sudo netstat -na | grep 1194
tcp6       0      0 2001:xxx:xxx:xxx.1194  *.*                    LISTEN
tcp4       0      0 10.xx.xx.xx.1194        *.*                    LISTEN
udp6       0      0 2001:xxx:xxx:xxx.1194  *.*
udp4       0      0 10.xx.xx.xx.1194        *.*
```

**Comment by fichtner** [comment]
Will fix the validation. This went in as part of a larger rework for OpenVPN dual stack usability. 

Shall we work on your IPv6 requests while at it?

**Comment by fichtner** [comment]
@sjorge actually, you have TCP and UDP running, can you switch those to TCP4 and UDP4?

**Comment by fichtner** [comment]
PS: If you see an error you should attach it. "I can just no longer edit any of them" is as good as guessing.

**Comment by sjorge** [comment]
@fichtner oops, I though I had pasted the message. Seems I did not. Update the original comment.
I cannot change the TCP and UDP to TCP4 and UDP4... given the validation is preventing me from saving anything. (Even though the config currently does work)

**Comment by fichtner** [comment]
Thanks, found it. Will fix in the morning.

**Comment by fichtner** [comment]
Try this:

    # opnsense-patch d210cbf 64c4a1f

But make sure it's TCP4 and UDP4, TCP and UDP means dual stack in OpenVPN and the code was adjusted to adhere to this situation.

**Comment by sjorge** [comment]
I still get the same error:
```
The following input errors were detected:
    The specified 'Local port' is in use. Please select another value
```

I tried to update the TCP one to TCP4 as you suggested.

**Comment by fichtner** [comment]
Works from here after applying both patches.

**Comment by fichtner** [comment]
Sorry, works for me even on 18.1.8. What version are you on?

**Comment by sjorge** [comment]
Should have mentioned that, not sure why I did not as I usually do: OPNsense 18.7.b_137-amd64

**Comment by fichtner** [comment]
Okay, please walk me through this:

Open ovpn-udp4, select "UDP4", save.

What error are you seeing?

**Comment by sjorge** [comment]
![screenshot](https://i.imgur.com/uUVVFNz.png)

**Comment by sjorge** [comment]
```
function openvpn_port_used($prot, $interface, $port, $curvpnid = 0)
{
    global $config;

    $af = null;

    if (strlen($prot) > 3) {
        $af = substr($prot, 3, 1);
        $prot = substr($prot, 0, 3);
    }

    if (isset($config['openvpn']['openvpn-server'])) {
        foreach ($config['openvpn']['openvpn-server'] as $settings) {
            if (isset($settings['disable'])) {
                continue;
            }

            if ($curvpnid != 0 && $curvpnid == $settings['vpnid']) {
                continue;
            }

            if ($port != $settings['local_port']) {
                continue;
            }

            if (strpos($settings['protocol'], $prot) === false) {
                continue;
            }

            if (!empty($af) && strlen($settings['protocol']) > 3 && strpos($settings['protocol'], $af) === false) {
                continue;
            }

            if ($interface != $settings['interface'] && $interface != 'any' && $settings['interface'] != 'any') {
                continue;
            }

            return $settings['vpnid'];
        }
    }

    if (isset($config['openvpn']['openvpn-client'])) {
        foreach ($config['openvpn']['openvpn-client'] as $settings) {
            if (isset($settings['disable'])) {
                continue;
            }

            if ($curvpnid != 0 && $curvpnid == $settings['vpnid']) {
                continue;
            }

            if ($port != $settings['local_port']) {
                continue;
            }

            if (strpos($settings['protocol'], $prot) === false) {
                continue;
            }

            if (!empty($af) && strlen($settings['protocol']) > 3 && strpos($settings['protocol'], $af) === false) {
                continue;
            }

            if ($interface != $settings['interface'] && $interface != 'any' && $settings['interface'] != 'any') {
                continue;
            }

            return $settings['vpnid'];
        }
    }

    return 0;
}
```

**Comment by fichtner** [comment]
Ok, just to be sure: do you have clients set up that could interfere?

**Comment by sjorge** [comment]
Nope, don’t have any clients. (Also never had those)

**Comment by fichtner** [comment]
Are you sure the GUI is on the same box that was patched via SSH? Still looks like it's working with your data. Confirmed twice.

**Comment by sjorge** [comment]
Yes, I only have one opnsense box at the moment.
I rechecked the files and tried again, I still get the error.

Maybe it was because mine were still at TCP and UDP for 2 of them, 
**I was able to get the config correctly by exporting (TCP->TCP4 and UDP->UDP4), editing the export and importing.** So importing config is at least working, I don't think that was every broken though.

![after_import](https://i.imgur.com/mVWJ6dj.png)

But when I try to edit and save the server without changes, I still get the error.

![edit_no_changes_save](https://i.imgur.com/2mJ99EA.png)
